File: lib/exception/f_exception_aggregate.dart
// Copyright 2023, the Freemework.ORG project authors. Please see the AUTHORS
// file for details. All rights reserved. Use of this source code is governed
// by a BSD-style license that can be found in the LICENSE file.

import "f_exception.dart" show FException;

class FExceptionAggregate extends FException {
  static void throwIfNeeded(final Iterable<FException> innerExceptions) {
    if (innerExceptions.isNotEmpty) {
      if (innerExceptions.length == 1) {
        throw innerExceptions.first;
      }
      throw FExceptionAggregate(innerExceptions);
    }
  }

  final List<FException> innerExceptions;
  factory FExceptionAggregate(final Iterable<FException> innerExceptions,
      [String? message]) {
    const defaultMessage = "One or more errors occurred.";
    return FExceptionAggregate._internal(
      message ?? defaultMessage,
      innerExceptions,
    );
  }

  FExceptionAggregate._internal(
      String? message, Iterable<FException> innerExceptions)
      : innerExceptions = List.unmodifiable(innerExceptions),
        super(
            message, innerExceptions.isNotEmpty ? innerExceptions.first : null);

  @override
  String toString() {
    final List<String> messages = [super.toString()];

    if (this.innerExceptions.isNotEmpty) {
      messages.addAll(this.innerExceptions.map((e) => e.toString()));
    }
    return messages.join("\n");
  }
}

File: src/exception/f_exception_aggregate.ts
import { EOL } from "os";

import { FException } from "./f_exception";

export class FExceptionAggregate extends FException {
	static throwIfNeeded(innerExceptions: ReadonlyArray<FException>): void {
		if (innerExceptions.length > 0) {
			if (innerExceptions.length == 1) {
				throw innerExceptions[0];
			}
			throw new FExceptionAggregate(innerExceptions);
		}
	}

	public readonly innerExceptions: ReadonlyArray<FException>;
	public constructor(innerExceptions: ReadonlyArray<FException>, message?: string) {
		let friendlyInnerException: FException | null;
		let friendlyInnerExceptions: Array<FException>;
		let friendlyMessage: string;
		if (innerExceptions.length > 0) {
			friendlyInnerExceptions = [...innerExceptions];
			friendlyInnerException = friendlyInnerExceptions.length > 0 ? friendlyInnerExceptions[0] : null;
		} else {
			friendlyInnerExceptions = [];
			friendlyInnerException = null;
		}
		if(message !== undefined) {
			friendlyMessage = message;
		} else {
			friendlyMessage = "One or more errors occurred.";
		}

		if (friendlyInnerException !== null) {
			super(friendlyMessage, friendlyInnerException);
		} else {
			super(friendlyMessage);
		}

		this.innerExceptions = Object.freeze(friendlyInnerExceptions);
	}

	/**
	 * @override
	 */
	public toString(): string {
		const  messages: Array<String> = [super.toString()];
	
		if (this.innerExceptions.length > 0) {
		  messages.push(...this.innerExceptions.map((e) => e.toString()));
		}
		return messages.join(EOL);
	  }
}
